### PR TITLE
cs/replay: check appHash for each block

### LINKED ISSUE #3460: replayBlocks checkAppHash should't wait the last block
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):
0d985ede28bd6937fa9d3613618e42cab6fc871c

**ABCI app** (name for built-in, URL for self-written if it's publicly available):

**Environment**:
- **OS** (e.g. from /etc/os-release):
- **Install tools**:
- **Others**:

**What happened**:
there are some code should be enhanced

**What you expected to happen**:
I will contribute code later
**Have you tried the latest version**: yes/no

**How to reproduce it** (as minimally and precisely as possible):

**Logs (paste a small part showing an error (< 10 lines) or link a pastebin, gist, etc. containing more of the log file)**:

**Config (you can paste only the changes you've made)**:

**node command runtime flags**:

**`/dump_consensus_state` output for consensus bugs**

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by melekes**
Fixes #3460

Not really needed, but it would detect if the app changed in a way it
shouldn't have.

* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: consensus/replay.go [hunk 1]
```diff
@@ -456,9 +462,26 @@ func (h *Handshaker) replayBlock(state sm.State, height int64, proxyApp proxy.Ap
 	return state, nil
 }
 
-func checkAppHash(state sm.State, appHash []byte) error {
-	if !bytes.Equal(state.AppHash, appHash) {
-		panic(fmt.Errorf("Tendermint state.AppHash does not match AppHash after replay. Got %X, expected %X", appHash, state.AppHash).Error())
```
Comment (Liamsi) [inline consensus/replay.go]: `panic(fmt.Errorf().Error())`, that was a weird way to panic 🙈

File: consensus/replay.go [hunk 2]
```diff
@@ -456,9 +462,26 @@ func (h *Handshaker) replayBlock(state sm.State, height int64, proxyApp proxy.Ap
...
+func checkAppHashAgainstBlock(appHash []byte, block *types.Block) error {
```
Comment (Liamsi) [inline consensus/replay.go]: Why does the methiod return an error? The only thing it does, is panic in an error case and return nil if everything seems right.

File: consensus/replay.go [hunk 3]
```diff
@@ -417,6 +417,12 @@ func (h *Handshaker) replayBlocks(state sm.State, proxyApp proxy.AppConns, appBl
 	for i := appBlockHeight + 1; i <= finalBlock; i++ {
 		h.logger.Info("Applying block", "height", i)
 		block := h.store.LoadBlock(i)
+
+		// Extra check to ensure the app was not changed in a way it shouldn't have.
+		if len(appHash) > 0 {
+			checkAppHashAgainstBlock(appHash, block)
```
Comment (Liamsi) [inline consensus/replay.go]: Why is the error never checked here? And why does the method never return an error (only panic).

File: consensus/replay.go [hunk 4]
```diff
@@ -456,9 +462,26 @@ func (h *Handshaker) replayBlock(state sm.State, height int64, proxyApp proxy.Ap
...
+		panic(fmt.Sprintf(`block.AppHash does not match AppHash after replay. Got %X, expected %X.
+
+Block: %v
+`,
+			appHash, block.AppHash, block))
+	}
+	return nil
+}
+
+func checkAppHashAgainstState(appHash []byte, state sm.State) error {
```
Comment (Liamsi) [inline consensus/replay.go]: Same question: This will never return an error.

File: consensus/replay.go (hunk 2 above)
Comment (melekes) [reply consensus/replay.go]: Good catch. No need to return an error

File: consensus/replay.go [hunk 5]
```diff
@@ -456,11 +465,26 @@ func (h *Handshaker) replayBlock(state sm.State, height int64, proxyApp proxy.Ap
 	return state, nil
 }
 
-func checkAppHash(state sm.State, appHash []byte) error {
-	if !bytes.Equal(state.AppHash, appHash) {
-		panic(fmt.Errorf("Tendermint state.AppHash does not match AppHash after replay. Got %X, expected %X", appHash, state.AppHash).Error())
+func checkAppHashAgainstBlock(appHash []byte, block *types.Block) {
```
Comment (Liamsi) [inline consensus/replay.go]: The name is slightly misleading. This isn't a simple check anymore as we only panic. Don't know what is the best name here to signal this but I think `assertXXX`works.  E.g.
`assertAppMatchesBlockHash`
`assertAppMatchesStateHash`

You could even combine this in a single method:
`assertAppHashEq(appHash []byte, otherHash []byte, panicMsg string)`

Comment (melekes) [reply consensus/replay.go]: Updated. Thanks!

**Comment by melekes** [comment]
I will probably take a lot of time, but I'll try. You must understand that requiring to write a test for each PR will slow down dev. speed to 0 until codebase is easily testable (then it will be fast again)

**Comment by melekes** [comment]
@Liamsi added a test case. PTAL

File: consensus/replay_test.go [hunk 6]
```diff
@@ -499,7 +637,7 @@ func makeBlockchainFromWAL(wal WAL) ([]*types.Block, []*types.Commit, error) {
 		return nil, nil, err
 	}
 	if !found {
-		return nil, nil, fmt.Errorf("WAL does not contain height %d.", 1)
+		return nil, nil, fmt.Errorf("WAL does not contain height %d", 1)
```
Comment (melekes) [inline consensus/replay_test.go]:
```suggestion
		return nil, nil, fmt.Errorf("WAL does not contain height %d", 0)
```

File: consensus/replay_test.go [hunk 7]
```diff
@@ -489,6 +495,138 @@ func buildTMStateFromChain(config *cfg.Config, stateDB dbm.DB, state sm.State, c
...
+		defer proxyApp.Stop()
+
+		assert.Panics(t, func() {
+			h := NewHandshaker(stateDB, state, store, genDoc)
+			h.Handshake(proxyApp)
+		})
+	}
+}
+
+func makeBlocks(n int, state *sm.State, privVal types.PrivValidator) []*types.Block {
```
Comment (melekes) [inline consensus/replay_test.go]: https://github.com/tendermint/tendermint/issues/3584

File: consensus/replay.go [hunk 8]
```diff
@@ -393,8 +395,7 @@ func (h *Handshaker) ReplayBlocks(
 
 	}
 
-	cmn.PanicSanity("Should never happen")
-	return nil, nil
+	panic("should never happen")
```
Comment (melekes) [inline consensus/replay.go]: explain why it should never happen?

File: consensus/replay_test.go [hunk 9]
```diff
@@ -489,6 +495,138 @@ func buildTMStateFromChain(config *cfg.Config, stateDB dbm.DB, state sm.State, c
...
+	//		- 0x03
+	{
+		app := &badApp{numBlocks: 3, allHashesAreWrong: true}
+		clientCreator := proxy.NewLocalClientCreator(app)
+		proxyApp := proxy.NewAppConns(clientCreator)
+		err := proxyApp.Start()
+		require.NoError(t, err)
+		defer proxyApp.Stop()
+
+		assert.Panics(t, func() {
```
Comment (melekes) [inline consensus/replay_test.go]: how to check panic msg using regexp?

**Pull request merged**